File: src/main/java/org/example/kalkulationsprogramm/controller/EmailTextTemplateController.java
package org.example.kalkulationsprogramm.controller;

import jakarta.validation.Valid;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.domain.Dokumenttyp;
import org.example.kalkulationsprogramm.domain.EmailTextTemplate;
import org.example.kalkulationsprogramm.domain.EmailTextTemplateKategorie;
import org.example.kalkulationsprogramm.dto.Email.EmailTextTemplateDto;
import org.example.kalkulationsprogramm.service.EmailTextTemplateKategorien;
import org.example.kalkulationsprogramm.service.EmailTextTemplateService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/email-textvorlagen")
@RequiredArgsConstructor
public class EmailTextTemplateController {

    // Single Source of Truth: Doc-Typen kommen aus dem Dokumenttyp-Enum.
    // ZEICHNUNG existiert nicht im Enum, wird aber als E-Mail-Vorlage benötigt
    // (Versand der ersten Entwurfs-PDF an den Kunden) und daher manuell ergänzt.
    // WEBSITE_ANFRAGE_BESTAETIGUNG ist die automatische Bestätigungsmail an
    // Leads aus dem Funnel der Marketing-Webseite — kein Geschaeftsdokument,
    // aber per E-Mail-Vorlage genauso editierbar.
    private static final List<Map<String, String>> DOKUMENT_TYPEN = buildDokumentTypen();

    private static List<Map<String, String>> buildDokumentTypen() {
        List<Map<String, String>> list = new ArrayList<>();
        for (Dokumenttyp d : Dokumenttyp.values()) {
            String label = switch (d) {
                case ANGEBOT -> "Anfrage / Angebot";
                default -> d.getLabel();
            };
            list.add(buildOption(d.name(), label));
        }
        list.add(buildOption("ZEICHNUNG", "Zeichnung / Entwurf"));
        list.add(buildOption("WEBSITE_ANFRAGE_BESTAETIGUNG", "Webseite — Anfragebestätigung"));
        return List.copyOf(list);
    }

    private static Map<String, String> buildOption(String value, String label) {
        // LinkedHashMap, damit die Reihenfolge der Keys im JSON deterministisch
        // ist — Map.of() ist intern unsortiert, was beim Snapshot-Vergleich in
        // den Controller-Tests irritierende Diff-Treffer erzeugen kann.
        Map<String, String> option = new LinkedHashMap<>();
        option.put("value", value);
        option.put("label", label);
        EmailTextTemplateKategorie kat = EmailTextTemplateKategorien.kategorieFuer(value);
        option.put("kategorie", kat.name());
        option.put("kategorieLabel", kat.getLabel());
        return option;
    }

    private static final List<Map<String, String>> PLACEHOLDERS = List.of(
            Map.of("token", "{{ANREDE}}", "label", "Anrede (z. B. Sehr geehrter Herr Müller)"),
            Map.of("token", "{{KUNDENNAME}}", "label", "Kundenname (Firma / Name des Kunden)"),
            Map.of("token", "{{ANSPRECHPARTNER}}", "label", "Ansprechpartner (Person beim Kunden)"),
            Map.of("token", "{{BAUVORHABEN}}", "label", "Bauvorhaben"),
            Map.of("token", "{{PROJEKTNUMMER}}", "label", "Projektnummer"),
            Map.of("token", "{{DOKUMENTNUMMER}}", "label", "Dokumentnummer (Rechnung/Auftrag/Anfrage)"),
            Map.of("token", "{{RECHNUNGSDATUM}}", "label", "Rechnungsdatum"),
            Map.of("token", "{{FAELLIGKEITSDATUM}}", "label", "Fälligkeitsdatum"),
            Map.of("token", "{{BETRAG}}", "label", "Betrag (formatiert)"),
            Map.of("token", "{{BENUTZER}}", "label", "Sachbearbeiter / Benutzer"),
            Map.of("token", "{{REVIEW_LINK}}", "label", "Google-Bewertungs-Link"),
            // Webseiten-Lead-Bestätigung
            Map.of("token", "{{NACHRICHT}}", "label", "Nachricht aus dem Webseiten-Funnel"),
            Map.of("token", "{{ANFRAGE_DATUM}}", "label", "Anfrage-Datum (Webseite)"),
            Map.of("token", "{{ANFRAGENUMMER}}", "label", "Anfrage-Nummer (Webseite)"));

    private final EmailTextTemplateService service;

    @GetMapping
    public ResponseEntity<List<EmailTextTemplateDto>> list() {
        List<EmailTextTemplateDto> result = service.list().stream()
                .map(EmailTextTemplateDto::fromEntity)
                .collect(Collectors.toList());
        return ResponseEntity.ok(result);
    }

    @GetMapping("/{id}")
    public ResponseEntity<EmailTextTemplateDto> get(@PathVariable Long id) {
        try {
            EmailTextTemplate entity = service.get(id);
            return ResponseEntity.ok(EmailTextTemplateDto.fromEntity(entity));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.notFound().build();
        }
    }

    @PostMapping
    public ResponseEntity<EmailTextTemplateDto> create(@Valid @RequestBody EmailTextTemplateDto dto) {
        EmailTextTemplate saved = service.create(dto);
        return ResponseEntity.ok(EmailTextTemplateDto.fromEntity(saved));
    }

    @PutMapping("/{id}")
    public ResponseEntity<EmailTextTemplateDto> update(@PathVariable Long id, @Valid @RequestBody EmailTextTemplateDto dto) {
        try {
            EmailTextTemplate saved = service.update(id, dto);
            return ResponseEntity.ok(EmailTextTemplateDto.fromEntity(saved));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.notFound().build();
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        service.delete(id);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/dokumenttypen")
    public ResponseEntity<List<Map<String, String>>> dokumenttypen() {
        return ResponseEntity.ok(DOKUMENT_TYPEN);
    }

    @GetMapping("/placeholders")
    public ResponseEntity<List<Map<String, String>>> placeholders() {
        return ResponseEntity.ok(PLACEHOLDERS);
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/controller/EmailTextTemplateController.kt
package org.example.kalkulationsprogramm.controller

import jakarta.validation.Valid
import org.example.kalkulationsprogramm.domain.Dokumenttyp
import org.example.kalkulationsprogramm.dto.Email.EmailTextTemplateDto
import org.example.kalkulationsprogramm.service.EmailTextTemplateKategorien
import org.example.kalkulationsprogramm.service.EmailTextTemplateService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/email-textvorlagen")
class EmailTextTemplateController(
    private val service: EmailTextTemplateService
) {
    @GetMapping
    fun list(): ResponseEntity<List<EmailTextTemplateDto>> =
        ResponseEntity.ok(service.list().map(EmailTextTemplateDto::fromEntity))

    @GetMapping("/{id}")
    fun get(@PathVariable id: Long): ResponseEntity<EmailTextTemplateDto> =
        try {
            ResponseEntity.ok(EmailTextTemplateDto.fromEntity(service.get(id)))
        } catch (_: IllegalArgumentException) {
            ResponseEntity.notFound().build()
        }

    @PostMapping
    fun create(@Valid @RequestBody dto: EmailTextTemplateDto): ResponseEntity<EmailTextTemplateDto> =
        ResponseEntity.ok(EmailTextTemplateDto.fromEntity(service.create(dto)))

    @PutMapping("/{id}")
    fun update(@PathVariable id: Long, @Valid @RequestBody dto: EmailTextTemplateDto): ResponseEntity<EmailTextTemplateDto> =
        try {
            ResponseEntity.ok(EmailTextTemplateDto.fromEntity(service.update(id, dto)))
        } catch (_: IllegalArgumentException) {
            ResponseEntity.notFound().build()
        }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Long): ResponseEntity<Void> {
        service.delete(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/dokumenttypen")
    fun dokumenttypen(): ResponseEntity<List<Map<String, String>>> = ResponseEntity.ok(DOKUMENT_TYPEN)

    @GetMapping("/placeholders")
    fun placeholders(): ResponseEntity<List<Map<String, String>>> = ResponseEntity.ok(PLACEHOLDERS)

    companion object {
        private val DOKUMENT_TYPEN: List<Map<String, String>> = buildDokumentTypen()

        private fun buildDokumentTypen(): List<Map<String, String>> {
            val list = ArrayList<Map<String, String>>()
            for (d in Dokumenttyp.entries) {
                val label = if (d == Dokumenttyp.ANGEBOT) "Anfrage / Angebot" else d.label
                list.add(buildOption(d.name, label))
            }
            list.add(buildOption("ZEICHNUNG", "Zeichnung / Entwurf"))
            list.add(buildOption("WEBSITE_ANFRAGE_BESTAETIGUNG", "Webseite - Anfragebestaetigung"))
            return list.toList()
        }

        private fun buildOption(value: String, label: String): Map<String, String> {
            val kat = EmailTextTemplateKategorien.kategorieFuer(value)
            val option = LinkedHashMap<String, String>()
            option["value"] = value
            option["label"] = label
            option["kategorie"] = kat.name
            option["kategorieLabel"] = kat.label
            return option
        }

        private val PLACEHOLDERS: List<Map<String, String>> = listOf(
            mapOf("token" to "{{ANREDE}}", "label" to "Anrede (z. B. Sehr geehrter Herr Mueller)"),
            mapOf("token" to "{{KUNDENNAME}}", "label" to "Kundenname (Firma / Name des Kunden)"),
            mapOf("token" to "{{ANSPRECHPARTNER}}", "label" to "Ansprechpartner (Person beim Kunden)"),
            mapOf("token" to "{{BAUVORHABEN}}", "label" to "Bauvorhaben"),
            mapOf("token" to "{{PROJEKTNUMMER}}", "label" to "Projektnummer"),
            mapOf("token" to "{{DOKUMENTNUMMER}}", "label" to "Dokumentnummer (Rechnung/Auftrag/Anfrage)"),
            mapOf("token" to "{{RECHNUNGSDATUM}}", "label" to "Rechnungsdatum"),
            mapOf("token" to "{{FAELLIGKEITSDATUM}}", "label" to "Faelligkeitsdatum"),
            mapOf("token" to "{{BETRAG}}", "label" to "Betrag (formatiert)"),
            mapOf("token" to "{{BENUTZER}}", "label" to "Sachbearbeiter / Benutzer"),
            mapOf("token" to "{{REVIEW_LINK}}", "label" to "Google-Bewertungs-Link"),
            mapOf("token" to "{{BANK}}", "label" to "Bank (aus Firmeneinstellungen)"),
            mapOf("token" to "{{IBAN}}", "label" to "IBAN (aus Firmeneinstellungen)"),
            mapOf("token" to "{{BIC}}", "label" to "BIC (aus Firmeneinstellungen)"),
            mapOf("token" to "{{NACHRICHT}}", "label" to "Nachricht aus dem Webseiten-Funnel"),
            mapOf("token" to "{{ANFRAGE_DATUM}}", "label" to "Anfrage-Datum (Webseite)"),
            mapOf("token" to "{{ANFRAGENUMMER}}", "label" to "Anfrage-Nummer (Webseite)")
        )
    }
}
